### PR TITLE
Code review and profiling

### PR DESCRIPTION
**Pull request by papiveron**
I modified the code while forcussing in this first pull request on the backup function and mecanism, for :
- Better management of source tables names for backup purposes, now one can pass a list of table sperated by the separator, or a wildcard handled like in unix shell as argument
- reduce for loops, queuing and if statements, for coding improvement
- suitable log format logger
- better management of s3 backup
- set default archive type to zip, and leave legacy choices of {zip, tar}
- put argument parsing in a separate function from the main function
- general review of the the whole backup part of the code

Look at my commits changes and messages and let me know if we agree.....

**Note :** As I said in my first commit, code improvement can still be done, I really appreciate the great job you did and I would like to contribute to your original project instead of working in this separated forked one. Please can you give commit rights after this review?

Cheers.

### CODE REVIEW

**Comment by papiveron** [comment]
I did many changes to the code, like:
- porting all the code to boto3
- better credentials management and aws cross account backup/restore enabled
- whole code review to correct several runtime errors

I'm still waiting for a review of my PR in order to let everyone benefit from the contributions I have made

Regards,
Eugene NG

**Comment by bchew** [comment]
@papiveron that's a lot of changes - can you add tests as that was what I have planned to do but have not got time for to prevent regression? Thanks.

File: dynamodump.py [hunk 1]
```diff
@@ -39,48 +37,68 @@
...
 LOCAL_REGION = "local"
 LOG_LEVEL = "INFO"
-DATA_DUMP = "dump"
+DATA_DUMP = "dynamodump"
+OUTPUT_DIR = "/tmp/dynamorestore/"
 RESTORE_WRITE_CAPACITY = 25
 THREAD_START_DELAY = 1  # seconds
 CURRENT_WORKING_DIR = os.getcwd()
-DEFAULT_PREFIX_SEPARATOR = "-"
+DEFAULT_PREFIX_SEPARATOR = ","
```
Comment (bchew) [inline dynamodump.py]: This is going to be a breaking change in default behaviour, I'll have to go through more in detail again

File: README.md [hunk 2]
```diff
@@ -3,18 +3,24 @@ dynamodump
...
-Simple backup and restore script for Amazon DynamoDB using boto to work similarly to mysqldump.
+Simple backup and restore script for Amazon DynamoDB using boto3 to work similarly to mysqldump.
 
 Suitable for DynamoDB usages of smaller data volume which do not warrant the usage of AWS Data Pipeline for backup/restores/empty.
 
 dynamodump supports local DynamoDB instances as well (tested with [dynalite](https://github.com/mhart/dynalite)).
 
+dynamodump support user credentials and IAM roles
+
+You can also use dynamodump for cross account backup/restore, thus you can run on one aws instance of an aws account to process dynamodb tables backup/restore/empty of another trusted aws account
```
Comment (bchew) [inline README.md]: Mind making `AWS` all caps? :)

File: README.md [hunk 3]
```diff
@@ -46,6 +53,11 @@ optional arguments:
                         AWS credentials file profile to use. Allows you to use
                         a profile instead of accessKey, secretKey
                         authentication
+  -u ASSUMEDACCOUNTID   AWS Account ID to use if the programm should assume an iam role from another account
```
Comment (bchew) [inline README.md]: Account -> account
programm -> dynamodump?
iam -> IAM

File: README.md [hunk 4]
```diff
@@ -46,6 +53,11 @@ optional arguments:
...
+  -n ASSUMEDROLENAME    Role name to be assumed
+  --sessionToken SESSIONTOKEN  
+                        Session tocken to use for aws credentials
```
Comment (bchew) [inline README.md]: `token` and `AWS`

File: dynamodump.py [hunk 5]
```diff
@@ -39,48 +37,68 @@
...
+        session = boto3.Session(profile_name=args.profile)
+        client = session.client(service, region_name=aws_region, endpoint_url=endpoint)
     else:
-        client = boto3.client(service, region_name=aws_region)
+        if (args.assumedAccountId):
+            if not args.assumedRoleName:
+                logging.error("You should specify an asumed role name for the provided asumed account id")
+                sys.exit(1)
+            else:
+                sts_client = boto3.client('sts')
```
Comment (bchew) [inline dynamodump.py]: Just to be consistent, use `"` rather than `'`

File: dynamodump.py [hunk 6]
```diff
@@ -39,48 +37,68 @@
...
+                assumedRoleObject = sts_client.assume_role(
+                    RoleArn="arn:aws:iam::%s:role/%s" % (args.assumedAccountId, args.assumedRoleName),
+                    RoleSessionName="AssumeRoleDynamoBAckup"
```
Comment (bchew) [inline dynamodump.py]: Backup rather than BAckup?

File: dynamodump.py [hunk 7]
```diff
@@ -39,48 +37,68 @@
...
+                )
+                credentials = assumedRoleObject['Credentials']
```
Comment (bchew) [inline dynamodump.py]: Quotes for the next few lines

File: dynamodump.py [hunk 8]
```diff
@@ -342,45 +392,45 @@ def change_prefix(source_table_name, source_wildcard, destination_wildcard, sepa
...
+            table_exists = True
             try:
-                conn.delete_table(table_name)
-            except boto.exception.JSONResponseError as e:
-                if e.body["__type"] == "com.amazonaws.dynamodb.v20120810#ResourceNotFoundException":
-                    table_exist = False
-                    logging.info(table_name + " table deleted!")
+                dynamo.delete_table(TableName=table_name)
+            except botocore.exceptions.ClientError as e:
+                if e.response['Error']['Code'] == "ResourceNotFoundException":
```
Comment (bchew) [inline dynamodump.py]: Quotes

**Comment by AlJohri** [comment]
hey @bchew, commenting here because Github issues isn't enabled. Can you enable `Issues` for this repository? https://help.github.com/articles/disabling-issues/

**Pull request closed without merging**